libfwupdplugin/fu-coswid-common.c

```diff
@@ -0,0 +1,375 @@
+ *
+ * Since: 1.9.17
+ **/
+gboolean
+fu_coswid_read_u8(cbor_item_t *item, guint8 *value, GError **error)
+{
+	g_return_val_if_fail(item != NULL, FALSE);
+	g_return_val_if_fail(value != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (!cbor_isa_uint(item)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_DATA,
+				    "value item is not a uint");
+		return FALSE;
+	}
+	*value = cbor_get_uint8(item);
+	return TRUE;
+}
+
+/**
+ * fu_coswid_read_u64:
+ * @item: a #cbor_item_t
+ * @value: read value
+ * @error: (nullable): optional return location for an error
+ *
+ * Reads a #guint64 tag value.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 1.9.17
+ **/
+gboolean
+fu_coswid_read_u64(cbor_item_t *item, guint64 *value, GError **error)
+{
+	g_return_val_if_fail(item != NULL, FALSE);
+	g_return_val_if_fail(value != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (!cbor_isa_uint(item)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_DATA,
+				    "value item is not a uint");
+		return FALSE;
+	}
+	*value = cbor_get_uint64(item);
+	return TRUE;
+}
+
+/**
+ * fu_coswid_write_tag_string:
+ * @item: a #cbor_item_t
+ * @tag: a #FuCoswidTag, e.g. %FU_COSWID_TAG_PAYLOAD
+ * @value: string
+ *
+ * Writes a string tag value.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 1.9.17
+ **/
+void
+fu_coswid_write_tag_string(cbor_item_t *item, FuCoswidTag tag, const gchar *value)
+{
+	g_autoptr(cbor_item_t) key = cbor_build_uint8(tag);
+	g_autoptr(cbor_item_t) val = cbor_build_string(value);
+	if (!cbor_map_add(item, (struct cbor_pair){.key = key, .value = val}))
+		g_critical("failed to push to indefinite map");
+}
+
+/**
+ * fu_coswid_write_tag_bytestring:
+ * @item: a #cbor_item_t
+ * @tag: a #FuCoswidTag, e.g. %FU_COSWID_TAG_PAYLOAD
+ * @buf: a buffer of data
+ * @bufsz: sizeof @buf
+ *
+ * Writes a bytestring tag value.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 1.9.17
+ **/
+void
+fu_coswid_write_tag_bytestring(cbor_item_t *item, FuCoswidTag tag, const guint8 *buf, gsize bufsz)
+{
+	g_autoptr(cbor_item_t) key = cbor_build_uint8(tag);
+	g_autoptr(cbor_item_t) val = cbor_build_bytestring((cbor_data)buf, bufsz);
+	if (!cbor_map_add(item, (struct cbor_pair){.key = key, .value = val}))
+		g_critical("failed to push to indefinite map");
+}
+
+/**
+ * fu_coswid_write_tag_bool:
+ * @item: a #cbor_item_t
+ * @tag: a #FuCoswidTag, e.g. %FU_COSWID_TAG_PAYLOAD
+ * @value: boolean
+ *
+ * Writes a #gboolean tag value.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 1.9.17
+ **/
+void
+fu_coswid_write_tag_bool(cbor_item_t *item, FuCoswidTag tag, gboolean value)
+{
+	g_autoptr(cbor_item_t) key = cbor_build_uint8(tag);
+	g_autoptr(cbor_item_t) val = cbor_build_bool(value);
+	if (!cbor_map_add(item, (struct cbor_pair){.key = key, .value = val}))
+		g_critical("failed to push to indefinite map");
```

Comment:
to make these easier to debug use unique strings for each critical

Reply:
good idea, fixed.